docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+
+* Entry flow:
+  * [Rough description]: the steps when a Tx enters the node are unmodified, except
+    * `CheckTx` (optionally) returns lane information (if TX valid)
+    * The lane information is used to decide to which transaction list the Tx will be added
+* Broadcast flow:
+  * Higher-priority lanes get to be gossiped first, and gossip within a lane is still in FIFO order.
+  * [Rough description]: we need to extend the loop in the broadcast goroutine. Several channels now, not just one
+    * To decide
+      * if we're going for (simpler) solution where with max 2 lanes,
+        the broadcast goroutine itself could manage the priority of lanes within the existing `select` at the end of the loop
+      * if we're going for (future-proof) solution where max lanes can be configurable,
+        we need to refactor method `TxWaitChan` (or the code that sends the data to that channel) to manage the lane priority there.
+        In this case, the broadcast goroutine will be mainly unmodified.
+* Reap flow
+  * TXs are reaped from higher-priority lanes first, in FIFO order.
+  * We go from the reap loop (currently just one), to probably a nested one
+    * the outer loop goes through all TX lists (remember, one per lane), in decreasing priority order
+    * the current `break` statements in the inner loop (limit of bytes or gas reached), should also break from the outer loop
+* Exit flow (Update, Re-Check). Update: unconditionally; Re-Check: only if App says TX is now invalid
+  * Transactions in higher-priority lanes are processed (updated) first.
+  * Depends on how the multiple TX lists are implemented (see discussion above)
+  * Update contains Re-check at the end
+  * Remember Update is done while holding the mempool lock (no new CheckTx calls... they have to wait)
+  * Changes:
+    * We could update the different TX lists in parallel
+    * `Update` method (implementing `Update` method in `Mempool` interface) is implemented by `CListMempool` so
+      * If we go for several `CListMempool`s, then we'll have to call `Update` on each of them with all the TXs every time
+        * or move `Update` method out of `CListMempool` (where?)
+      * Also, the way we will do `ReCheck` would be complicated if we had N `Update` calls
+        * We will do `ReCheck` lane-by-lane, in FIFO order within a lane
+      * So, this seems to make us lean toward: one `CListMempool` containing N `txs` lists
+* Each lane informs consensus when there are txs available.
+  *  `TxsAvailable` is in our laundry list
+
+ADR: Core part of detailed design. Bullets will likely become subsections
+
+### Who defines lanes and priorities
+
+The list of lanes and their priorities: are consensus parameters? are defined by the app?
+* Discussion on _where_ the lane configuration is set up. Three options
+  1. `config.toml` / `app.toml`
+  2. ConsensusParams
+  3. Hardcoded in the application. Info passed to CometBFT during handshake
+* Outcome of discussion
+  * 1. vs. 2.
```

Comment:
for me this doesn't render properly (`a. vs. 2.`)

Reply:
Yes, should be `i.`

Reply:
Thanks for pointing out. We'll double-check when the text becomes more mature.